executor/aggregate.go

```diff
@@ -45,39 +105,490 @@ type HashAggExec struct {
+	finalWorkers     []HashAggFinalWorker
+	defaultVal       *chunk.Chunk
+	// isChildExecReturnEmpty indicates whether the child executor only returns an empty input.
+	isChildExecReturnEmpty bool
+}
+
+// HashAggInput indicates the input of hash agg exec.
+type HashAggInput struct {
+	chk        *chunk.Chunk
+	giveBackCh chan<- *chunk.Chunk
+}
+
+// HashAggIntermData indicates the intermediate data of aggregation execution.
+type HashAggIntermData struct {
+	groupKeys   [][]byte
+	cursor      int
+	groupCtxMap aggCtxsMapper
+}
+
+// ToRows converts HashAggInterData to DatumRows.
+func (d *HashAggIntermData) ToRows(sc *stmtctx.StatementContext, rows []types.DatumRow, aggFuncs []aggregation.Aggregation, maxChunkSize int) (_ []types.DatumRow, reachEnd bool) {
+	if len(rows) == maxChunkSize {
+		return rows, false
+	}
+	for ; d.cursor < len(d.groupKeys); d.cursor++ {
+		groupKey := d.groupKeys[d.cursor]
+		row := make(types.DatumRow, 0, len(aggFuncs)*2)
+		aggCtxs := d.groupCtxMap[string(groupKey)]
+		for i, f := range aggFuncs {
+			for _, d := range f.GetPartialResult(aggCtxs[i]) {
+				row = append(row, d)
+			}
+		}
+		// Append groupKey as the last element.
+		row = append(row, types.NewBytesDatum(groupKey))
+		rows = append(rows, row)
+	}
+	return rows, true
 }
 
 // Close implements the Executor Close interface.
 func (e *HashAggExec) Close() error {
-	if err := e.baseExecutor.Close(); err != nil {
-		return errors.Trace(err)
+	if e.doesUnparallelExec {
+		e.groupMap = nil
+		e.groupIterator = nil
+		e.aggCtxsMap = nil
+	} else {
+		// `Close` may be called after `Open` without calling `Next` in test.
+		if !e.prepared {
+			close(e.inputCh)
+			for _, ch := range e.partialOutputChs {
+				close(ch)
+			}
+			close(e.finalOutputCh)
+		}
+		close(e.finishCh)
+		for _, ch := range e.partialOutputChs {
+			for range ch {
+			}
+		}
+		for range e.finalOutputCh {
+		}
 	}
-	e.groupMap = nil
-	e.groupIterator = nil
-	e.aggCtxsMap = nil
-	return nil
+	return errors.Trace(e.baseExecutor.Close())
 }
 
 // Open implements the Executor Open interface.
 func (e *HashAggExec) Open(ctx context.Context) error {
 	if err := e.baseExecutor.Open(ctx); err != nil {
 		return errors.Trace(err)
 	}
-	e.executed = false
+	e.prepared = false
+
+	if e.doesUnparallelExec {
+		e.initForUnparallelExec()
+		return nil
+	}
+	e.initForParallelExec()
+	return nil
+}
+
+func (e *HashAggExec) initForUnparallelExec() {
 	e.groupMap = mvmap.NewMVMap()
 	e.groupIterator = e.groupMap.NewIterator()
 	e.aggCtxsMap = make(aggCtxsMapper, 0)
 	e.mutableRow = chunk.MutRowFromTypes(e.retTypes())
 	e.rowBuffer = make([]types.Datum, 0, e.Schema().Len())
 	e.groupKey = make([]byte, 0, 8)
 	e.groupVals = make([][]byte, 0, 8)
+}
+
+func (e *HashAggExec) initForParallelExec() {
+	sessionVars := e.ctx.GetSessionVars()
+	finalConcurrency := sessionVars.HashAggFinalConcurrency
+	partialConcurrency := sessionVars.HashAggPartialConcurrency
+	e.isChildExecReturnEmpty = true
```

Comment:
Why set this to true?

Reply:
Since we do not know whether child executor will return values when executing.
So, when HashAgg gets the first batch of data, this value can be set to false.